### PR TITLE
Get user transactions from database instead of redux state

### LINKED ISSUE #2171: Move transaction list to Apollo entirely
## Description

This is a refactoring task. We would like to move the transaction list to GraphQL entirely. Right now it's a bit of a weird mix of redux state (for real time) and Apollo state (for past transactions). The goal is to remove the transaction list from Redux and add realtime capabilities to Apollo.

For that, WebSocket and subscriptions support is needed for Apollo, GraphQL and Amplify, as well as the auth proxy.
Furthermore we need subscription handlers for new transactions and updating existing ones.
Then we need to add optimistic updates for quick response times that people are already used to from the Redux implementation.

### Design

Not applicable

### This issue is complete when...

- [ ] No transactions sit on the Redux state anymore
- [ ] All transactions live exclusively on Apollo
- [ ] The behavior is exactly the same for users of the CDapp

### PR DESCRIPTION
**Pull request by chmanie**
## Description

Sorry for another big one :(

This PR moves the transaction state to Apollo, from redux. It _does not_ remove sagas but it enables a path forward to refactor this gradually. All sagas _should_ still work, except that the transaction state is read from the Apollo cache / the database as opposed to redux.

Redux as a whole also wasn't removed yet, as some of the sagas depend on other parts of the state, still (like gas prices estimation). This can be done in a follow-up PR.

Another thing that isn't in scope is the fix of the transactions list paging (didn't work before), for that I'm going to open yet another PR.

I used optimistic UI updates for the transaction updates so that changes should be reflected as fast as before. 

## Testing

Basically everything should work as before. Please create some transactions and/or bigger transaction groups to test this.

## Diffs

**Changes** 🏗

* All redux transaction state now lives in Apollo

**Deletions** ⚰️

* Lots of transaction related redux state management boilerplate

Resolves #2171

### CODE REVIEW

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Great effort went into this PR, nice work @chmanie!

I spotted a potential regression when a tx fails:
- on `master`, it expands the error details:

https://github.com/JoinColony/colonyCDapp/assets/112586815/910c67be-c498-4196-80e2-375974923fb9


- whereas on this branch it doesn't happen:

https://github.com/JoinColony/colonyCDapp/assets/112586815/9c61ed44-efba-4e8e-868e-af8993ecef29

**Comment by chmanie** [comment]
> Great effort went into this PR, nice work @chmanie!
> 
> I spotted a potential regression when a tx fails:
> 
>     * on `master`, it expands the error details:
> 
> 
> Screen.Recording.2024-05-10.at.13.52.05.mov
> 
>     * whereas on this branch it doesn't happen:
> 
> 
> Screen.Recording.2024-05-10.at.13.55.52.mov

Thanks for the review @jakubcolony! I'll take care of it asap.

**Comment by chmanie** [comment]
@jakubcolony can you take another look maybe? Thank you!

**Review by jakubcolony — CHANGES_REQUESTED** [review]
@chmanie I noticed userhub styles break when going tablet/mobile:

image

The colony creation flow seems broken:

https://github.com/JoinColony/colonyCDapp/assets/112586815/a24b6363-ed22-4652-92f3-d9f12dd4b66c

File: src/state/transactionState.ts [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+    fetchPolicy,
+  });
+
+  if (!data.getTransaction) {
+    throw new Error('Transaction not found.');
+  }
+  return convertTransactionType(data.getTransaction);
+};
+
+export const fetchTransaction = async ({ id }) => {
```
Comment (bassgeta) [inline src/state/transactionState.ts:227]: do we need both this and `getTransaction`?

File: src/state/transactionState.ts [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+    title: JSON.stringify(group.title) || null,
+    titleValues: JSON.stringify(group.titleValues) || null,
+  };
+
+  const txCreatedAt = createdAt.toISOString();
+  const txParams = JSON.stringify(params || []);
+
+  const apollo = getContext(ContextModule.ApolloClient);
+
+  const input = {
```
Comment (bassgeta) [inline src/state/transactionState.ts:281]: maybe I'd type this just for code completion

**Review by bassgeta — DISMISSED** [review]
D A M N, this is a big one, but it feels like a nice improvement.
I think the only "limitations" are due to redux-saga, so can't do much about it :shrug: 
I haven't gotten the time to test it proper thorough, but a smoke test goes through fine, code changes make everything simpler and more readable.
Let's get this one in :ship:

File: src/state/transactionState.ts (hunk 1 above)
Comment (chmanie) [reply src/state/transactionState.ts:227]: That is a very good point, let me look into this.

File: src/state/transactionState.ts (hunk 2 above)
Comment (chmanie) [reply src/state/transactionState.ts:281]: Gotcha.

**Comment by chmanie** [comment]
Ok, addressed your comments, @bassgeta.

Also @jakubcolony, please check again at your hearts content.

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Hey @chmanie I've just tried the creation flow and it's looking much better, however I noticed the transaction group in the userhub is always expanded and it doesn't react when you click on the arrow:

https://github.com/JoinColony/colonyCDapp/assets/112586815/43e48ac5-38f0-4778-9027-841e64c4a93b

**Comment by chmanie** [comment]
> Hey @chmanie I've just tried the creation flow and it's looking much better, however I noticed the transaction group in the userhub is always expanded and it doesn't react when you click on the arrow:
> 
> 
> 
> https://github.com/JoinColony/colonyCDapp/assets/112586815/43e48ac5-38f0-4778-9027-841e64c4a93b
> 
> 
> 
> 
> 
> 
> 
> 

Ah, I thought this was the expected behavior! I'll look at the current state again and copy that :)

File: amplify/backend/api/colonycdapp/schema.graphql [hunk 3]
```diff
@@ -1601,7 +1601,7 @@ type Transaction @model {
   """
   The group to which the transaction belongs, if any
```
[the remark this answers is not in the document]
Comment (chmanie) [reply amplify/backend/api/colonycdapp/schema.graphql:1602]: Good catch!

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Great work on this Chris! 🎉 I tested a bunch of different actions and motions and mostly looking good! I was able to create a new colony no issue.

Screenshot 2024-07-04 at 15 57 22

I did find this one nice bug with the 'pending' button. If a transaction doesn't go through, then the 'pending' button will never disappear even after a refresh.

https://github.com/JoinColony/colonyCDapp/assets/34915414/19187930-f664-43ae-a5ca-ab7cf8a4ba62

(I was still able to complete other actions after which lead to this nice 3/2 behaviour).

Seems to be an issue with locking an expenditure:

https://github.com/JoinColony/colonyCDapp/assets/34915414/091a28e1-b5a4-45ec-9265-fe99d6eca6ba

I'm not having the locking expenditure issue on master.

**Comment by chmanie** [comment]
OK, @iamsamgibbs, I think this is it. May thou please runneth it and check again?

Serious Sam

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Great job with this Chris, the previous issue is resolved.

Screenshot 2024-07-16 at 13 51 50

I am having an issue whenever the userhub is forced to open by a failed transaction, it is impossible to close the userhub.

https://github.com/user-attachments/assets/25602038-084b-4b8b-b70b-d62c053304e9

And this is a personal opinion thing, but I think any transaction that only has 1 action should just show `Pending` rather than `Pending 0/1`

Screenshot 2024-07-16 at 14 15 47

**Comment by chmanie** [comment]
Oh my, this goes on and on, doesn't it. Thanks for catching this.
I'll also make the change for the one pending tx if @arrenv is also fine with it

**Review by davecreaser — DISMISSED** [review]
Wow, okay this was a huge amount of work, nice one @chmanie 🍻 

A lot of things look a lot cleaner and I think this makes it easier to understand too. I've gone through the code as throughly as I could for such a large change, and there's nothing major that stands out to me, just a couple of tiny nitpicks, but they won't stop me from approving this.

Also spent the last half an hour testing lots of things locally on this branch - colony creation, passed transactions, pending transactions, failed transactions, retrying transactions, cancelling transactions etc. Everything seems to be working as expected, and I've had a quick look at the history of comments / changes on this PR and checked that they're still good too. I won't spam the comments with lots of screenshots of the testing now, but at least here's some nice proof that failing transactions and cancelling transactions works:

Screenshot 2024-07-25 at 11 13 19
Screenshot 2024-07-25 at 11 13 24

All in all, a great job, and thanks for being patient with the long review process. Apologies I didn't get to it sooner!

Let's get it in before it grows anymore 🤙

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Great work on this Chris! Thanks for making the `Pending 0/1` change, looks a lot nicer now.

I tested installing extensions, creating a colony, creating a bunch of actions, etc. The only issue I hit was when trying to stake a motion:

https://github.com/user-attachments/assets/6a99547d-b80a-4a14-b3a3-cd56b10716e6

It wouldn't work unless I activated my tokens first:

https://github.com/user-attachments/assets/9f17a9b6-62cd-4cc2-bb12-a78c028aa43d

I've tested it on master and can confirm it isn't an issue there.

**Comment by chmanie** [comment]
> Great work on this Chris! Thanks for making the `Pending 0/1` change, looks a lot nicer now.
> 
> I tested installing extensions, creating a colony, creating a bunch of actions, etc. The only issue I hit was when trying to stake a motion:
> Screen.Recording.2024-07-25.at.16.29.01.mov
> 
> It wouldn't work unless I activated my tokens first:
> Screen.Recording.2024-07-25.at.16.29.31.mov
> 
> I've tested it on master and can confirm it isn't an issue there.

That's so very odd. Good work catching this, I'll test to see what the issue is here.

File: src/components/v5/shared/TxButton/TxButton.tsx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      setShowPending(true);
+    } else if (
+      groupState === TransactionGroupStatus.Done &&
+      prevGroupState === TransactionGroupStatus.Pending
+    ) {
+      setShowPending(false);
+      setShowCompleted(true);
+      setTimeout(() => {
+        setShowCompleted(false);
+      }, 5000);
```
[the remark this answers is not in the document]
Comment (chmanie) [reply src/components/v5/shared/TxButton/TxButton.tsx:65]: Very good catch!

File: src/components/common/Extensions/UserHub/partials/TransactionsTab/partials/hooks.tsx [hunk 5]
```diff
@@ -7,26 +6,18 @@ import { TransactionStatus } from '~gql';
 import {
   transactionEstimateGas,
   transactionCancel,
-  transactionRetry,
 } from '~redux/actionCreators/index.ts';
 import { type TransactionId } from '~redux/immutable/index.ts';
 import Toast from '~shared/Extensions/Toast/index.ts';
-import { type ExtendedClientType } from '~types/transactions.ts';
+
+import { transactionRetry } from '../../../../../../../state/transactionState.ts';
```
[the remark this answers is not in the document]
Comment (chmanie) [reply src/components/common/Extensions/UserHub/partials/TransactionsTab/partials/hooks.tsx:13]: I always prefer lint rules over manual inspection in PRs. If you want to add that, you have my vote :)

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
The staking issue is fixed, but the "completed" pill isn't disappearing after a transaction:

Screenshot 2024-07-26 at 10 26 42

**Comment by chmanie** [comment]
> The staking issue is fixed, but the "completed" pill isn't disappearing after a transaction:
> Screenshot 2024-07-26 at 10 26 42

Argh, apologies, that's me not testing it because I thought it was a no-brainer fix. Well, it appears that I indeed had no brain when fixing it.

**Review by iamsamgibbs — COMMENTED** [review]
Nice job Chris! Completed pill issue is resolved:

https://github.com/user-attachments/assets/373a8a22-2408-4534-a5b9-ec95d14d753d

I have found one issue, but it might also exist on master.

I created a mint tokens action, but rejected the first transaction. I then retried it from the userhub, this time accepting the transaction, but the "Claim pending transactions" part never triggered. This left the "Pending 1/2" pill spinning forever, even after a refresh, and there is no way to retry it from the userhub.

Screenshot 2024-07-29 at 09 28 53

**Comment by iamsamgibbs** [comment]
Just tested this on master - it does behave the same, except the "pending" pill disappears after a refresh. I think the change here to show the "pending" pill after a refresh makes sense as there are still pending transactions, but I wouldn't be surprised if we find more stuck "pending" pills in other areas of the app.

**Review by iamsamgibbs — APPROVED** [review]

**Comment by chmanie** [comment]
Oh yeah that's definitely something that needs to be fixed but can't be in the current state. As a matter of fact, the transaction API refactor that I'm working on will fix this issue.

**Review by davecreaser — APPROVED** [review]
Nice one Chris, everything is still looking good to me. Played around a little locally again to be sure and couldn't spot any regressions.

Let's go! 🎉
